src/hotspot/cpu/x86/assembler_x86.hpp

```diff
@@ -789,14 +789,32 @@ class Assembler : public AbstractAssembler  {
   void vex_prefix(bool vex_r, bool vex_b, bool vex_x, int nds_enc, VexSimdPrefix pre, VexOpcode opc);
 
   void evex_prefix(bool vex_r, bool vex_b, bool vex_x, bool evex_v, bool evex_r, bool evex_b,
-                       bool eevex_x, int nds_enc, VexSimdPrefix pre, VexOpcode opc);
+                       bool eevex_x, int nds_enc, VexSimdPrefix pre, VexOpcode opc, bool no_flags = false);
+
+  void vex_prefix_ndd(Address adr, int ndd_enc, int xreg_enc, VexSimdPrefix pre, VexOpcode opc, InstructionAttr *attributes, bool no_flags = false) {
+    vex_prefix(adr, ndd_enc, xreg_enc, pre, opc, attributes, /* nds_is_ndd */ true , /* force_evex */ true, no_flags);
+  }
+
+  void vex_prefix_nf(Address adr, int ndd_enc, int xreg_enc, VexSimdPrefix pre, VexOpcode opc, InstructionAttr *attributes, bool no_flags = false) {
+    vex_prefix(adr, ndd_enc, xreg_enc, pre, opc, attributes, /* nds_is_ndd */ false , /* force_evex */ true, no_flags);
+  }
 
   void vex_prefix(Address adr, int nds_enc, int xreg_enc, VexSimdPrefix pre, VexOpcode opc,
-                  InstructionAttr *attributes);
+                  InstructionAttr *attributes, bool nds_is_ndd = false, bool force_evex = false, bool no_flags = false);
 
   int  vex_prefix_and_encode(int dst_enc, int nds_enc, int src_enc,
                              VexSimdPrefix pre, VexOpcode opc,
-                             InstructionAttr *attributes, bool src_is_gpr = false);
+                             InstructionAttr *attributes, bool src_is_gpr = false, bool nds_is_ndd = false, bool force_evex = false, bool no_flags = false);
+
+  int  vex_prefix_and_encode_ndd(int dst_enc, int nds_enc, int src_enc, VexSimdPrefix pre, VexOpcode opc,
```

Comment:
The second argument (nds_enc) carries the ndd encoding.  I kept the original parameter names since they are used in the implementation, which now serves both the ndd and nds features.  The added nds_is_ndd flag distinguishes the two uses.  

The two encoding functions (vex_prefix and vex_prefix_and_encode) were fairly complex to start with.  Changes to them to get ndd and nf features are pretty simple, so reusing them and keeping the existing parameter names in the implementation seemed a good idea.  Using these same parameter names in the header took away the need to translate names if going between the header and impl.